lambdas/access_counts/index.py

```diff
@@ -0,0 +1,161 @@
+            args.update(
+                NextToken=next_token
+            )
+
+
+def write_counts_csv(fd, data, field_names):
+    writer = csv.writer(fd)
+    writer.writerow(field_names)
+    for key, value in data.items():
+        row = list(key)
+        row.append(json.dumps(value))
+        writer.writerow(row)
+
+
+def handler(event, context):
+    named_packages = defaultdict(list)  # (bucket, hash) -> [(name, version), ...]
+
+    # (eventname, bucket, key) -> {date: count, ...}
+    object_access_counts = defaultdict(lambda: defaultdict(int))
+
+    # (eventname, bucket, package_name) -> {date: count, ...}
+    package_access_counts = defaultdict(lambda: defaultdict(int))
+
+    # (eventname, bucket, package_name, package_version, package_hash) -> {date: count, ...}
+    package_version_access_counts = defaultdict(lambda: defaultdict(int))
+
+    named_packages_execution_id = run_query(NAMED_PACKAGES_QUERY_ID)
+    access_counts_execution_id = run_query(OBJECT_ACCESS_COUNTS_QUERY_ID)
+
+    named_packages_key = '%s/%s.csv' % (QUERY_TEMP_DIR, named_packages_execution_id)
+    access_counts_key = '%s/%s.csv' % (QUERY_TEMP_DIR, access_counts_execution_id)
+
+    wait_for_query(named_packages_execution_id)
+
+    for bucket, name, version, hash in get_query_results(named_packages_execution_id):
+        named_packages[(bucket, hash)].append((name, version))
+
+    wait_for_query(access_counts_execution_id)
+
+    for eventname, date, bucket, key, count_str in get_query_results(access_counts_execution_id):
+        count = int(count_str)
+
+        object_access_counts[(eventname, bucket, key)][date] += count
+
+        parts = key.split('/')
+        if len(parts) == 3 and parts[0] == '.quilt' and parts[1] == 'packages':
+            package_hash = parts[2]
+            for package_name, package_version in named_packages.get((bucket, package_hash), []):
+                package_access_counts[(eventname, bucket, package_name)][date] += count
+                package_version_access_counts[(eventname, bucket, package_name, package_version, package_hash)][date] += count
+
+    for key in [named_packages_key, access_counts_key]:
+        s3.delete_object(Bucket=QUERY_RESULT_BUCKET, Key=key)
+        s3.delete_object(Bucket=QUERY_RESULT_BUCKET, Key=key + '.metadata')
+
+    outputs = [
+        ('Objects.csv', object_access_counts, ['eventname', 'bucket', 'key', 'counts']),
+        ('Packages.csv', package_access_counts, ['eventname', 'bucket', 'name', 'counts']),
+        ('PackageVersions.csv', package_version_access_counts, ['eventname', 'bucket', 'name', 'version', 'hash', 'counts']),
+    ]
+
+    for name, data, header in outputs:
+        with TemporaryFile('w+') as file_obj:
+            write_counts_csv(file_obj, data, header)
+            file_obj.seek(0)
+            s3.put_object(
+                Body=file_obj.buffer.raw,
+                Bucket=QUERY_RESULT_BUCKET,
+                Key=ACCESS_COUNTS_OUTPUT_DIR + '/' + name,
```

Comment:
not `os.path.join`? i realize this is an s3 path

Reply:
Exactly. `os.path.join` is for OS paths.